sphinx/ext/intersphinx/_load.py

```diff
@@ -138,36 +139,38 @@ def load_mappings(app: Sphinx) -> None:
     intersphinx_cache: dict[InventoryURI, InventoryCacheEntry] = inventories.cache
     intersphinx_mapping: IntersphinxMapping = app.config.intersphinx_mapping
 
+    expected_uris = {uri for uri, _invs in app.config.intersphinx_mapping.values()}
+
+    # If the current cache contains some (project, uri) pair
+    # say ("foo", "foo.com") and if the new intersphinx dict
+    # contains the pair ("foo", "bar.com"), we need to remove
+    # the ("foo", "foo.com") entry and use ("foo", "bar.com").
+    for uri in frozenset(intersphinx_cache):
+        if intersphinx_cache[uri][0] not in intersphinx_mapping or uri not in expected_uris:
+            # remove a cached inventory if the latter is no more used by intersphinx
+            del intersphinx_cache[uri]
+
     with concurrent.futures.ThreadPoolExecutor() as pool:
-        futures = []
-        for name, (uri, invs) in intersphinx_mapping.values():
-            futures.append(pool.submit(
-                fetch_inventory_group, name, uri, invs, intersphinx_cache, app, now,
-            ))
+        futures = [
+            pool.submit(fetch_inventory_group, name, uri, invs, intersphinx_cache, app, now)
+            for name, (uri, invs) in app.config.intersphinx_mapping.values()
+        ]
         updated = [f.result() for f in concurrent.futures.as_completed(futures)]
 
     if any(updated):
+        # clear the local inventories
         inventories.clear()
 
         # Duplicate values in different inventories will shadow each
-        # other; which one will override which can vary between builds
-        # since they are specified using an unordered dict.  To make
-        # it more consistent, we sort the named inventories and then
-        # add the unnamed inventories last.  This means that the
-        # unnamed inventories will shadow the named ones but the named
-        # ones can still be accessed when the name is specified.
-        named_vals = []
-        unnamed_vals = []
-        for name, _expiry, invdata in intersphinx_cache.values():
-            if name:
-                named_vals.append((name, invdata))
-            else:
-                unnamed_vals.append((name, invdata))
-        for name, invdata in sorted(named_vals) + unnamed_vals:
-            if name:
-                inventories.named_inventory[name] = invdata
-            for type, objects in invdata.items():
-                inventories.main_inventory.setdefault(type, {}).update(objects)
+        # other and which one will override which varies between builds.
+        #
+        # We can however order the cache by (NAME, EXPIRY) for reproducibility.
+        by_name_and_time = itemgetter(0, 1)  # 0: name, 1: expiry
+        cache_values = sorted(intersphinx_cache.values(), key=by_name_and_time)
+        for name, _expiry, invdata in cache_values:
```

Comment:
There was an additional check for empty `name` values here previously, removed in bd423ff4e5f1bf121d50692689b99fda1f10e343 - is that because we trust that the cache has been rebuilt already?

Reply:
Ok - currently I'm thinking about projects that might use the new format `intersphinx_mapping` but chose empty-string names; I don't find many, but there are a few: https://github.com/search?q=path%3Aconf.py+intersphinx_mapping+%2F%5E%5B+%5D*%5B%27%22%5D%5B%27%22%5D%3A+%5C%28%2F&type=code

If+when they upgrade, what happens first: an attempt to load the cache, or validation of the `intersphinx_mapping` config value?

Reply:
Validation of the mapping happens first (it's as soon as the configuration is created).

Reply:
(The loading happens when the builder is created)

Reply:
Ok.  It sorta seems to me that empty-string cache entries could currently still be loaded, and then would shadow entries from named inventories (not ideal, probably?), if I read this comment correctly:

https://github.com/sphinx-doc/sphinx/blob/9e3f4521db37f3216e1b256f6227ed852f54b879/sphinx/ext/intersphinx/_load.py#L152-L158

Reply:
The code that iterates over the entries also seems to confirm that -- in particular the fact that `unnamed_vals` are appended, so will be iterated _after_ the sorted named entries, and then may clobber dictionary entries.

https://github.com/sphinx-doc/sphinx/blob/9e3f4521db37f3216e1b256f6227ed852f54b879/sphinx/ext/intersphinx/_load.py#L166-L168

Reply:
(this data is sourced from cache -- not `intersphinx_mapping` config)

Reply:
>  It sorta seems to me that empty-string cache entries could currently still be loaded, and then would shadow entries from named inventories (not ideal, probably?),

Actually, there are no empty-string cache entries at all. 

1. If you run Sphinx with an older cache, then this cache will be wiped out because of the environment version having changed. So you will never be able to have those unnamed versions.
2. If you try to create a cache entry with an empty name, the validation will complain and you will never be able to load the inventories (the function will not even be called).

So, even if the current release is still able to load empty named inventories, in production you won't encounter any of them (at least, you shouldn't).

Reply:
In that case can we remove the code that handles `unnamed_vals` from the cache (or add a warning if it is now code that should not be reached)?

Reply:
Ack, yep you did.  I was comparing the wrong diff/version and thought that the code was still there.